File: android/src/main/java/com/uncompress/presentation/components/Uncompress.kt
package com.uncompress.presentation.components

import com.uncompress.data.intefaces.IExtractUseCase
import com.uncompress.data.intefaces.IIsProtectedUseCase

data class Uncompress(private val extractUseCase: IExtractUseCase, private val isProtectedUseCase: IIsProtectedUseCase) {

  @Throws(Throwable::class)
  fun extract(
    filePath: String,
    destination: String,
    override: Boolean = true,
    password: String? = null
  ) {
    extractUseCase.run(filePath, destination, override, password)
  }

  @Throws(Throwable::class)
  fun isProtected(filePath: String): Boolean {
    return isProtected(filePath)
  }

}

File: ios/Presentation/Components/Uncompress.swift
//
//  Uncompress.swift
//  Uncompress
//
//  Created by Adriano Souza Costa on 05/05/21.
//

import Foundation

public final class Uncompress {
    
    // MARK: - Private Properties
    
    private let extractUseCase: ExtractUseCaseProtocol
    private let isProctedUseCase: IsProtectedUseCaseProtocol
    
    // MARK: - Public Methods
    
    public init(
        extractUseCase: ExtractUseCaseProtocol,
        isProctedUseCase: IsProtectedUseCaseProtocol
    ) {
        self.extractUseCase = extractUseCase
        self.isProctedUseCase = isProctedUseCase
    }
    
    public func extract(
        _ filePath: String,
        to destination: String,
        overwrite: Bool = true,
        password: String? = nil
    ) throws {
        try extractUseCase.run(
            filePath,
            to: destination,
            overwrite: overwrite,
            password: password
        )
    }
    
    public func isProtected(_ filePath: String) throws -> Bool {
        return try isProctedUseCase.run(filePath)
    }
    
}
